library/x509_crl.c

```diff
@@ -294,7 +294,7 @@ int mbedtls_x509_crl_parse_der( mbedtls_x509_crl *chain,
     /*
      * Copy raw DER-encoded CRL
      */
-    if( ( p = mbedtls_calloc( 1, buflen ) ) == NULL )
+    if( buflen != 0 && ( ( p = mbedtls_calloc( 1, buflen ) ) == NULL ) )
         return( MBEDTLS_ERR_X509_ALLOC_FAILED );
 
     memcpy( p, buf, buflen );
```

Comment:
The test suite `test_suite_x509parse` fails in the clang-asan build (https://jenkins-internal.mbed.com/job/mbedtls-pr/425/execution/node/192/log/). I can reproduce this locally and identified the culprit as the change above.

```
X509 CRL ASN1 (Incorrect first tag) ............................... /home/gilpes01/z/dev/mbedtls/gate/library/x509_crl.c:300:13: runtime error: null pointer passed as argument 1, which is declared to never be null
/usr/include/string.h:43:28: note: nonnull attribute specified here
SUMMARY: AddressSanitizer: undefined-behavior /home/gilpes01/z/dev/mbedtls/gate/library/x509_crl.c:300:13 in 
```

This test case calls `mbedtls_x509_crl_parse` with `buflen = 0`. This in turn calls `mbedtls_x509_crl_parse_der` with `buflen = 0`, and this PR changed the way the function behaves in this case.

It's good that we don't rely on the behavior of `calloc` with a size of 0, but the way this was fixed, we're now calling `memcpy` with a size of 0. Calling `memcpy` with a null pointer and a size of 0 probably works on most platforms, but it's technically undefined behavior, and Clang-Asan rejects it.

I think that `memcpy(NULL, buf, 0)` is poor practice even if it works, and a rejection from Asan is not good, so I am not merging this PR.

Technically this is a pre-existing bug, since we returned `MBEDTLS_ERR_X509_ALLOC_FAILED` instead of `MBEDTLS_ERR_X509_INVALID_FORMAT` on platforms where `calloc(…, 0)` returns NULL. But I think the best solution would be to add
```
    if( buflen == 0 )
        return(  MBEDTLS_ERR_X509_INVALID_FORMAT );
```

near the top of this function.